Add Support for OpenAI-compatible Embeddings API

FabianScheidt opened the pull request:
**What this PR does / why we need it**:

This PR extends the KServe protocol to support an [OpenAI-compatible endpoint for text embeddings](https://platform.openai.com/docs/api-reference/embeddings). It adds a simple implementation to the HuggingFace serving runtime, which essentially wraps the existing implementation for the open inference protocol. 

**Which issue(s) this PR fixes**
Fixes #4025 

**Type of changes**

- [x] New feature (non-breaking change which adds functionality)

**Feature/Issue validation/testing**:

Embeddings models can be served using the HuggingFace Serving Runtime. E.g., like this

```bash
python -m huggingfaceserver \
    --model_name=embedding \
    --model_id=mixedbread-ai/mxbai-embed-large-v1 \
    --task=text_embedding
```

Previously, it was only possible to call the embedding model using the open inference protocol. We're retaining this functionality:

```bash
curl -H "Content-Type: application/json" \
    "http://localhost:8080/v1/models/embedding:predict" \
    -d '{"instances": ["Hello World"]}' | jq
```

Now, we're introducing an OpenAI-compatible endpoint:

```bash
curl http://localhost:8080/openai/v1/embeddings \
  -H "Content-Type: application/json" \
  -d '{
    "input": ["Hello World"],
    "model": "embedding",
    "encoding_format": "float"
  }' | jq
```

As a result, we're able to call the endpoint with [OpenAI's python library](https://pypi.org/project/openai/):

```python
from openai import OpenAI

client = OpenAI(base_url='http://localhost:8080/openai/v1', api_key='no-api-key-needed')
res = client.embeddings.create(model="embedding", input=["Hello World"])
print(res)
```

This yields the following output:

```
CreateEmbeddingResponse(data=[Embedding(embedding=[0.015021147206425667, 0.020543351769447327, ...], index=0, object='embedding')], model='embedding', object='list', usage=Usage(prompt_tokens=4, total_tokens=4))
```

**Special notes for your reviewer**:

This PR includes lots of boilerplate changes to the OpenAI data plane. In essence, we're separating the `OpenAIModel` class into `OpenAICompletionModel` and `OpenAIEmbeddingModel`. We're using `OpenAICompletionModel` in all the places where we previously had `OpenAIModel`. 

The actual implementation of the OpenAI-support in the HuggingFace serving runtime is simple: We're inheriting from `OpenAIEmbeddingModel` and implement a method to wrap the existing implementation.

**Checklist**:

- [x] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

I've added tests for the model registration. Not sure what other tests would be beneficial. Happy to make changes if you have suggestions. In case this is merged, we should add a remark to the [API endpoints section of the documentation](https://kserve.github.io/website/latest/modelserving/v1beta1/llm/huggingface/#api-endpoints).

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Add Support for OpenAI-compatible Embeddings API in HuggingFace Serving Runtime
```

FabianScheidt committed:
Add boilerplate setup for embeddings api

Signed-off-by: Fabian Scheidt <fabian.scheidt@accenture.com>

FabianScheidt committed:
Integrate call to model with embedding request

Signed-off-by: Fabian Scheidt <fabian.scheidt@accenture.com>

mginfn commented:
Hi @FabianScheidt, thanks for this work!
In my case, I'm implementing a fully-hosted QA solution and I need OpenAI-compatible Embeddings API.
Are there any plans to review/approve this PR?
Thanks

yuzisun commented:
> Hi @FabianScheidt, thanks for this work! In my case, I'm implementing a fully-hosted QA solution and I need OpenAI-compatible Embeddings API. Are there any plans to review/approve this PR? Thanks

@Datta0 @gavrissh Can you help review this?

gavrissh commented on python/huggingfaceserver/huggingfaceserver/encoder_model.py:
```diff
@@ -340,3 +349,60 @@ def _log_request(self, request_id: str, prompt: list[str]) -> None:
+    async def create_embedding(self, request: EmbeddingRequest) -> Embedding:
+        params = request.params
+        if params.input is None:
+            raise ValueError("input is required")
```
Let's use `OpenAIError()`, to return OpenAI Error Response json

gavrissh commented on python/huggingfaceserver/huggingfaceserver/encoder_model.py:
```diff
@@ -340,3 +349,60 @@ def _log_request(self, request_id: str, prompt: list[str]) -> None:
+
+        # The OpenAI documentation allows the input of token lists instead of strings. As the tokenization is specific
+        # to the model, it is most likely different from the ones used by OpenAI (e.g., tiktoken). Libraries like
+        # Llangchain attempt to determine the proper tokenization based on the model name and will fall back to the
```
nit: Langchain?

FabianScheidt replied:
If you're referring to the typo in the library name, that's fixed now :) If you're referring to the comment itself: It was the second library I experimented with after OpenAI's official library and I guess it's fairly common to use. I want to make sure that other people don't stumble into weird errors...

gavrissh commented on python/huggingfaceserver/huggingfaceserver/encoder_model.py:
```diff
@@ -340,3 +349,60 @@ def _log_request(self, request_id: str, prompt: list[str]) -> None:
+                and isinstance(params.input[0][0], int)
+            )
+            if input_is_int or input_is_list_of_int:
+                raise NotImplementedError("Input of token lists is not supported.")
```
Use `OpenAIError()`

gavrissh commented on python/kserve/kserve/protocol/rest/openai/endpoints.py:
```diff
@@ -137,6 +139,41 @@ async def stream_results() -> AsyncGenerator[str, None]:
+
+        Args:
+            raw_request (Request): fastapi request object,
+            model_name (str): Model name.
```
model_name is not an arg here?

FabianScheidt replied:
This was actually outdated across all doc strings in the file. I now have updated all of them.

gavrissh commented on python/kserve/kserve/protocol/dataplane.py:
```diff
@@ -432,8 +432,8 @@ async def infer(
         model = await self.get_model(model_name)
-        if isinstance(model, OpenAIModel):
-            error_msg = f"Model {model_name} is of type OpenAIModel. It does not support the infer method."
+        if isinstance(model, OpenAICompletionModel):
```
Is this change required?

FabianScheidt replied:
Yes, otherwise we'd break the current endpoints for v1 and v2 for no good reason.

FabianScheidt committed:
Adjust error responses to be consistent with openai

Signed-off-by: Fabian Scheidt <fabian.scheidt@accenture.com>

FabianScheidt committed:
Adjust doc strings to match implementation

Signed-off-by: Fabian Scheidt <fabian.scheidt@accenture.com>

Datta0 commented on python/huggingfaceserver/huggingfaceserver/encoder_model.py:
```diff
@@ -340,3 +351,72 @@ def _log_request(self, request_id: str, prompt: list[str]) -> None:
+        # accepting the mismatch, we rather raise an exception.
+        if isinstance(params.input, list) and len(params.input) > 0:
+            input_is_int = isinstance(params.input[0], int)
+            input_is_list_of_int = (
```
From what I understand, [`List[List[Any]]`](https://github.com/kserve/kserve/blob/654d314f15af5a4ed5fd6114dbd2be1cfdc6d5a0/python/kserve/kserve/protocol/rest/openai/types/openapi.py#L773) is valid for this. So if we want to enforce that only string or list of strings are input, shouldn't we directly check the same instead of checking if its not `int/List[int]` ?
Something like {'input':[[{"1":"2"}],[{"3":"4"}]] } would be valid according to this check right?
And it'd throw error down the line at tokenization but we want to avoid that ig?

FabianScheidt replied:
Good catch! I've adjusted the type validation to make use of pydantic, so it's more expressive. I have also adjusted the initial error message to match the one from the OpenAI API.

Datta0 commented on python/huggingfaceserver/huggingfaceserver/encoder_model.py:
```diff
@@ -340,3 +351,72 @@ def _log_request(self, request_id: str, prompt: list[str]) -> None:
+                embedding_out[i] = embedding_base64.decode("ascii")
+
+        # The autogenerated OpenAPI specification doesn't seem to include the base64 encoding option. Add it.
+        class EmbeddingObjectWithBase64Support(EmbeddingObject):
```
NIT: Can we move it to [__init__.py](https://github.com/kserve/kserve/blob/master/python/kserve/kserve/protocol/rest/openai/types/__init__.py) or [openai_model.py](https://github.com/kserve/kserve/blob/master/python/kserve/kserve/protocol/rest/openai/openai_model.py)

FabianScheidt replied:
I have moved it to \_\_init.py__. As a result, we also no longer need to lengthy name, but can simply override the existing class name. Hope that's alright with you :)

FabianScheidt committed:
Update openai input validation to use pydantic

Signed-off-by: Fabian Scheidt <fabian.scheidt@accenture.com>

FabianScheidt committed:
Refactor embedding object type

Signed-off-by: Fabian Scheidt <fabian.scheidt@accenture.com>

Datta0 reviewed (approved):
LGTM!
Thanks for these changes :)

gavrissh reviewed (approved):
/lgtm
Thank you!

yuzisun commented:
@FabianScheidt Can you help fix the CI ? Let's also add the embedding unit test to `python/kserve/test/test_openai.py` and the e2e test https://github.com/kserve/kserve/blob/master/test/e2e/predictor/test_huggingface.py#L294

FabianScheidt committed:
Add e2e test for OpenAI embedding endpoint

Signed-off-by: Fabian Scheidt <fabian.scheidt@accenture.com>

FabianScheidt committed:
Add unit test for OpenAI embeddings

Signed-off-by: Fabian Scheidt <fabian.scheidt@accenture.com>

FabianScheidt commented:
> @FabianScheidt Can you help fix the CI ? Let's also add the embedding unit test to `python/kserve/test/test_openai.py` and the e2e test https://github.com/kserve/kserve/blob/master/test/e2e/predictor/test_huggingface.py#L294

Sure, done!

- The failing CI was caused by a newer language feature for type hints, which I've adjusted.
- I have created a unit test in a separate file and renamed the existing one to differentiate between *completion* and *embedding*. That way, I don't need to rename all occurrences of the mocked model class.
- I have added an e2e test which validates both output formats *float* and *base64*.

FabianScheidt commented:
Sorry, I didn't realize that there are numerical inaccuracies even if we're using the base64 encoding option. I've updated the e2e test to decode the response and match against the approximation. Now the tests should pass.

gavrissh commented:
/rerun-all

yuzisun commented:
@FabianScheidt Thanks for the awesome contributions! 
/lgtm
/approve
